### PR TITLE
[mlir][Vector] Fix an assertion on failing cast in vector-transfer-flatten-patterns

### PR DESCRIPTION
**Pull request by bviyer**
When the result is not a vectorType, there is an assert. This patch will do the check 
and bail when the result is not a VectorType.

### CODE REVIEW

**Comment by joker-eph** [comment]
Thanks for the fix!

Can you add a test that shows how the assert is triggered?

**Comment by bviyer** [comment]
@joker-eph I added a test that will fail without the fix.

**Comment by banach-space** [comment]
> @joker-eph I added a test that will fail without the fix.

I don't hit the assert when `LLVM_ENABLE_ASSERTIONS` are ON. How do you trigger it?

**Comment by joker-eph** [comment]
I'm confused, the test does not match the code change: did you try running the test at HEAD and confirm it is crashing?

**Comment by bviyer** [comment]
> I'm confused, the test does not match the code change: did you try running the test at HEAD and confirm it is crashing?

The test was correct one but I accidentally put in the wrong file. Very sorry about the confusion.

Now, if you revert this check-in (afc2639af07bff07d887cfa38dd6eb872762d9cb) and then run `check-mlir` it would show 1 crash, the newly added test. With this above-mentioned check-in the whole thing should pass.

**Review by dcaballe — APPROVED** [review]
LG, thanks!

**Comment by bviyer** [comment]
Hi @joker-eph and @banach-space, Please let me know if you have any more questions. I would like to land the patch by Monday (March 25th 2024) evening if I don't hear any objections.

**Review by banach-space — CHANGES_REQUESTED** [review]
> I would like to land the patch by Monday (March 25th 2024) evening if I don't hear any objections.

Thanks for the update. I've made a suggestion how to improve the test - I'd like that to be addressed before this is merged.

File: mlir/test/Dialect/Vector/linearize.mlir
Comment (bviyer) [reply mlir/test/Dialect/Vector/linearize.mlir]: As you suggested, I removed the comment, nenamed the function more appropriately, added a comment about what this function tests and add check lines. Does this sound good to you?

File: mlir/test/Dialect/Vector/vector-transfer-flatten.mlir
Comment (bviyer) [reply mlir/test/Dialect/Vector/vector-transfer-flatten.mlir]: Removed!

File: mlir/test/Dialect/Vector/linearize.mlir
Comment (banach-space) [inline mlir/test/Dialect/Vector/linearize.mlir:100]: This is unrelated to this PR, but repeating labels like this is unnecessary noise that should be avoided. I will prepare a PR to trim this.

File: mlir/test/Dialect/Vector/linearize.mlir
Comment (banach-space) [inline mlir/test/Dialect/Vector/linearize.mlir:107]: This is much better, thanks! Now, this example doesn't linearise, because:

1. inputs are already linear (this is tested elsewhere),
2. inputs are tensors (this is unique).

In order to avoid mixing 1. and 2. in one test, could you make the tensors 2D? 

This is my final comment, I'll approve once this comment is addressed. Thanks!

Comment (bviyer) [reply mlir/test/Dialect/Vector/linearize.mlir]: I made the test a 2D tensor (2x2xf32) and added a check for a 2d multiplication.

**Review by banach-space — APPROVED** [review]
LGTM, thanks for addressing my comments!

**Pull request merged**
